src/Page.php

```diff
@@ -0,0 +1,169 @@
+{
+    const FINAL_PAGE_TOKEN = "";
+
+    private $parameters;
+    private $callable;
+    private $pageStreamingDescriptor;
+
+    private $pageToken;
+
+    private $response;
+
+    public function __construct($params, $callable, $pageStreamingDescriptor) {
+        if (empty($params) || !is_object($params[0])) {
+            throw new InvalidArgumentException('First argument must be a request object.');
+        }
+        $this->parameters = $params;
+        $this->callable = $callable;
+        $this->pageStreamingDescriptor = $pageStreamingDescriptor;
+
+        $requestPageTokenField = $this->pageStreamingDescriptor->getRequestPageTokenField();
+        $this->pageToken = $params[0]->$requestPageTokenField;
+
+        // Make gRPC call eagerly
+        $this->response = call_user_func_array($this->callable, $this->parameters);
+    }
+
+    /**
+     * Returns true if there are more pages that can be retrieved from the
+     * API.
+     */
+    public function hasNextPage() {
+        return strcmp($this->getNextPageToken(), Page::FINAL_PAGE_TOKEN) != 0;
+    }
+
+    /**
+     * Returns the next page token from the response.
+     */
+    public function getNextPageToken() {
+        $responsePageTokenField = $this->pageStreamingDescriptor->getResponsePageTokenField();
+        return $this->getResponseObject()->$responsePageTokenField;
+    }
+
+    /**
+     * Retrieves the next Page object using the next page token.
+     */
+    public function getNextPage($pageSize = null) {
+        if (!$this->hasNextPage()) {
+            throw new ValidationException(
+                'Could not complete getNextPage operation: ' .
+                'there are no more pages to retrieve.');
+        }
+
+        $newRequest = clone $this->getRequestObject();
+
+        $requestPageTokenField = $this->pageStreamingDescriptor->getRequestPageTokenField();
+        $newRequest->$requestPageTokenField = $this->getNextPageToken();
+
+        if (isset($pageSize)) {
+            $newRequest->setPageSize($pageSize);
+        }
+
+        $nextParameters = [$newRequest, $this->parameters[1], $this->parameters[2]];
+
+        return new Page($nextParameters, $this->callable, $this->pageStreamingDescriptor);
+    }
+
+    /**
+     * Return the number of elements in the response.
+     */
+    public function getPageElementCount() {
+        $resourceField = $this->pageStreamingDescriptor->getResourceField();
+        return count($this->getResponseObject()->$resourceField);
+    }
+
+    /**
+     * Return an iterator over the elements in the response.
+     */
+    public function iteratePageElements() {
+        $resourceField = $this->pageStreamingDescriptor->getResourceField();
+        foreach ($this->getResponseObject()->$resourceField as $element) {
+            yield $element;
+        }
+    }
+
+    /**
+     * Return an iterator over the elements in the response.
+     */
+    public function getIterator() {
```

Comment:
It seems a bit redundant to me to have both getIterator() and iteratePageElements()